### ISSUE TITLE
Nuxt build failing due to Terser

### ISSUE THREAD

**Issue by stepbeek** [nuxt/nuxt.js#4931]
### Version

[v2.4.2](https://github.com/nuxt.js/releases/tag/v2.4.2)

### Reproduction link

[https://github.com/nuxt/create-nuxt-app](https://github.com/nuxt/create-nuxt-app)

### Steps to reproduce

Create a new project using npx and try to build. 

### What is expected ?

A successful build.

### What is actually happening?

Build fails with error:

ERROR in baa9a7368c0a13e03198.js from Terser
TypeError: Cannot read property 'minify' of undefined
    at minify (/usr/src/app/node_modules/terser-webpack-plugin/dist/minify.js:175:23)

### Additional comments?

It looks like this is an upstream issue. After some time spent trying to figure out why my build was failing and not seeing any issues here, I figured I would open one with the link to what I believe the cause to be: https://github.com/webpack-contrib/terser-webpack-plugin/issues/66

Hopefully some other dev is saved some debug time!

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8593">#c8593</a>)</em></sub></div>

**Comment by simllll** [comment]
backtrace:

ERROR in 70fe8ac2e7db24c37f38.js from Terser
TypeError: Cannot read property 'minify' of undefined
    at minify (/home/simon/Dev/Web/hokify-pwa/node_modules/terser-webpack-plugin/dist/minify.js:175:23)
    at module.exports (/home/simon/Dev/Web/hokify-pwa/node_modules/terser-webpack-plugin/dist/worker.js:13:40)
    at handle (/home/simon/Dev/Web/hokify-pwa/node_modules/worker-farm/lib/child/index.js:44:8)
    at process.<anonymous> (/home/simon/Dev/Web/hokify-pwa/node_modules/worker-farm/lib/child/index.js:51:3)
    at process.emit (events.js:188:13)
    at emit (internal/child_process.js:828:12)
    at process.internalTickCallback (internal/process/next_tick.js:72:19)


terser is accessed via ".default".. which is not defined though. If you change the line to (/home/simon/Dev/Web/hokify-pwa/node_modules/terser-webpack-plugin/dist/minify.js:175)
```
 const {
    error,
    map,
    code,
    warnings
  } = _terser.minify({
    [file]: input
  }, terserOptions);
```

thinks work again..

**Comment by marcoeh** [comment]
Adding
```json
"resolutions": {
  "terser": "3.14.1"
}
```
to `package.json` helped me to successfully `$ yarn build` until this is fixed.

**Comment by stepbeek** [comment]
Similarly, hardcoding a dependency in my package.json worked for npm.

"dependencies":{
  ...
  "terser": "3.14.1"
}

**Comment by kevinmarrec** [comment]
@stepbeek Can you use `yarn why terser` and paste the result here ? It seems that something installed `terser@3.16.0` and made somehow Nuxt use it instead of `terser@3.14.1`.

**Comment by mauxtin** [comment]
Related: https://github.com/vuejs/vue-cli/issues/3407

**Comment by kevinmarrec** [comment]
Thanks @mauxtin. Fix on his way : https://github.com/terser-js/terser/pull/254

But the issue it not around Nuxt dependencies, it's around a module/option selected through `create-nuxt-app` which depends on `terser` and probably updated it to last minor version under a patch version.

But I can't easily find which one, I would need someone with the issue to do `yarn why terser`.

**Comment by marcoeh** [comment]
@kevinmarrec `"nuxt#@nuxt#webpack#terser-webpack-plugin"` depends on it

**Comment by bdrtsk** [comment]
I have the same issue!

**Comment by InstanceOfMichael** [comment]
```
$ npm ls terser
textsort@1.0.0 /home/iom/code/app
└─┬ nuxt@2.4.2
  └─┬ @nuxt/webpack@2.4.2
    └─┬ terser-webpack-plugin@1.2.1
      └── terser@3.16.0 
```
```
$ yarn why terser
yarn why v0.27.5
[1/4] Why do we have the module "terser"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
info This module exists because "nuxt#@nuxt/webpack#terser-webpack-plugin" depends on it.
info Disk size without dependencies: "3.85MB"
info Disk size with unique dependencies: "4.84MB"
info Disk size with transitive dependencies: "4.85MB"
info Number of shared dependencies: 4
Done in 0.71s.
```

**Comment by bdrtsk** [comment]
Still no fix, what a joke :smile: 

Have a productive week, guys

**Comment by bdrtsk** [comment]
@marcoeh you have something to tell me? Or you think it's normal to break millions of apps by not providing `default export` and not fixing it for days (it's will take like 15 minutes)? Everything not working, from `create-react-app` to Nuxt. ALl because dude forget to make `default export`.

**Comment by marcoeh** [comment]
https://github.com/terser-js/terser/pull/254#issuecomment-460008665

**Comment by robcresswell** [comment]
@bdrtsk If it will take 15 minutes, raise a PR yourself.

**Comment by robcresswell** [comment]
Even better, it's already in progress :smiley:

**Comment by bdrtsk** [comment]
@robcresswell yeah, PR :smile: It doesn't merged, that's the whole point. It's in progress for a 2 days already.

**Comment by pavlushacoder** [comment]
After watching Vuemastery tutorial about Nut I am trying to run `npx create-nuxt-app` and it's not working. Is this because of that? Can't understand (my first try of Nuxt)

**Comment by HoraceKeung** [comment]
Sorry, I should have said it better, Nuxt does not directly depend on Terser, but Nuxt depends on something that depends on Terser. I had the same issue on Saturday, spent 2-3 hours looking for a fix, then I see Terser released 3.16.0 on Saturday (https://www.npmjs.com/package/terser), after reading a bit more, I tried adding `"terser": "3.14.1"` in my dependencies, and it worked.

**Comment by HoraceKeung** [comment]
I found this commit in Terser, https://github.com/terser-js/terser/commit/620e1ece0bbc2e5adf331f6f05cb822d3f70aa80

**Comment by bdrtsk** [comment]
@HoraceKeung nice found!

And look at this.

https://twitter.com/swyx/status/1091959223406084096

![Uploading Screenshot from 2019-02-04 12-07-50.png…]()

It was 14 hours ago, so package owner knows about issue, but **he just like it**
Yeah, tell me more about personal life of maintainers.

**Comment by HoraceKeung** [comment]
The problem is he should have make it a major release i.e. make it version 4.0.0 then I believe none of these will happen.

**Comment by manniL** [comment]
Okay, we got it. Now calm down everybody 😉 

A few things you guys should consider:

1. There are things called **lockfiles**. If you haven't committed them to your Git repo, you now know **why you should** (read more [here](https://github.com/webpack-contrib/terser-webpack-plugin/issues/66#issuecomment-460161816))
2. For all people using Nuxt the first time or on a new project (especially @supersamka and @pavlushacoder), that's an unlucky start 😦 
3. There are **workarounds** for that issue known
4. Maintainers break things. That happens. They don't do that intentionally
5. No Maintainers **owes you anything**. We all put a lot of time into our projects, be it Nuxt, Webpack or whatever. And you all can use them **for free**
6. If you want that they owe you something, make a deal (e.g. a support contract or whatever) with them 🤷‍♂️
7. "Mimimi, 2 days without a fix how can you do that?!?!" will achieve nothing besides that maintainers get sick of such comments, the "community" (though the majority didn't do anything) and open source in general. **STOP IT!**

# TL;DR

* Use lockfiles
* Be excellent to each other
* Use **known workarounds** (look [here](https://github.com/vuejs/vue-cli/issues/3407#issuecomment-459985313) for example)

**Comment by manniL** [comment]
https://github.com/terser-js/terser/issues/251 / https://www.npmjs.com/package/terser/v/3.16.1

Terser `v3.16.1` resolves the issue

**Comment by bdrtsk** [comment]
@manniL so, "Mimimi" helped after all?

**Comment by kiloc** [comment]
"resolutions": {
    "@nuxt/**/terser": "3.14.1"
  }

try this ? i resolved
